File: 10.正则表达式匹配.c
/*
 * @lc app=leetcode.cn id=10 lang=c
 *
 * [10] 正则表达式匹配
 */
bool isMatch(char *s, char *p)
{
    bool dp[strlen(s) + 1][strlen(p) + 1];
    dp[strlen(s)][strlen(p)] = true;
    for (int i = strlen(s); i >= 0; i--)
    {
        for (int j = strlen(p) - 1; j >= 0; j--)
        {
            bool match = i < strlen(s) && s[i] == p[j] || p[j] == '.';
            if (j + 1 < strlen(p) && p[j + 1] == '*')
            {
                dp[i][j] = dp[i][j + 2] || (match && dp[i + 1][j]);
            }
            else
            {
                dp[i][j] = match && dp[i + 1][j + 1];
            }
        }
    }
    return dp[0][0];
}

File: 10.正则表达式匹配.java
/*
 * @lc app=leetcode.cn id=10 lang=java
 *
 * [10] 正则表达式匹配
 */
class Solution {
    public boolean isMatch(String s, String p) {
        if (p.isEmpty()) {
            return s.isEmpty();
        }
        boolean first_match = (!s.isEmpty() && (s.charAt(0) == p.charAt(0) || p.charAt(0) == '.'));
        if (p.length() >= 2 && p.charAt(1) == '*') {
            return (isMatch(s, p.substring(2)) || (first_match && isMatch(s.substring(1), p)));
        } else {
            return first_match && isMatch(s.substring(1), p.substring(1));
        }
    }
}
